File: 이준영/백준/Gold/16236. 아기 상어/아기 상어.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.StringTokenizer;

public class Main {
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		int N = Integer.parseInt(st.nextToken());
		int[][] arr = new int[N][N];
		int y = 0;
		int x = 0;
		int time = 0;
		int size = 2;
		int count = 0;
		int[] dy = { -1, 0, 0, 1 };
		int[] dx = { 0, -1, 1, 0 };

		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < N; j++) {
				arr[i][j] = Integer.parseInt(st.nextToken());
				if (arr[i][j] == 9) {
					y = i;
					x = j;
					arr[i][j] = 0;
				}
			}
		}

		while (true) {
			ArrayList<int[]> lst = new ArrayList<>();
			boolean[][] visited = new boolean[N][N];
			lst.add(new int[] { y, x, 0 });
			visited[y][x] = true;
			boolean check = false;
			while (true) {
				ArrayList<int[]> lst2 = new ArrayList<>();
				ArrayList<int[]> lst3 = new ArrayList<>();
				for (int[] tmp : lst) {
					int t_y = tmp[0];
					int t_x = tmp[1];
					
					for (int k = 0; k < 4; k++) {
						int ny = t_y + dy[k];
						int nx = t_x + dx[k];
						if (0 <= nx && nx < N && 0 <= ny && ny < N && !visited[ny][nx]) {
							visited[ny][nx] = true;
							if (arr[ny][nx] == 0 || arr[ny][nx] == size) {
								lst2.add(new int[] { ny, nx, tmp[2] + 1 });
							} else if (arr[ny][nx] < size) {
								lst3.add(new int[] { ny, nx, tmp[2] + 1 });
							}
						}
					}
				}
				if (!lst3.isEmpty()) {
					Collections.sort(lst3, (a, b) -> ((a[0] - b[0]) * 20 + a[1] - b[1]));
					int[] tmp = lst3.get(0);
					time = time + tmp[2];
					y = tmp[0];
					x = tmp[1];
					count+=1;
					arr[y][x] = 0;
					if(count==size) {
						size+=1;
						count=0;
					}
					break;
				} else if(!lst2.isEmpty()){
					lst = lst2;
				} else {
					check = true;
					break;
				}
			}
			if(check) {
				break;
			}
		}
		System.out.println(time);

	}
}

File: 이준영/백준/Gold/16236. 아기 상어/아기 상어.py
from collections import deque

n = int(input())
arr = [list(map(int, input().split())) for _ in range(n)]
dy = [0, 0, 1, -1]
dx = [1, -1, 0, 0]
time = 0
r_y = -1
r_x = -1
for i in range(n):
    for j in range(n):
        if arr[i][j] == 9:
            r_y = i
            r_x = j
            arr[i][j] = 0
            break
    if r_y != -1:
        break

r_size = 2
r_exp = 0
time = 0
while True:
    visited = [[0] * n for _ in range(n)]
    lst = []
    visited[r_y][r_x] = 1
    dq = deque([(r_y, r_x)])
    t_time = 1
    while dq:
        l = len(dq)
        for _ in range(l):
            y, x = dq.popleft()
            for k in range(4):
                ny = y + dy[k]
                nx = x + dx[k]
                if 0 <= ny < n and 0 <= nx < n and visited[ny][nx] == 0:
                    visited[ny][nx] = 1
                    if arr[ny][nx] > r_size:
                        continue
                    else:
                        dq.append((ny, nx))
                        if 0 < arr[ny][nx] < r_size:
                            lst.append((ny, nx))
                        else:
                            dq.append((ny, nx))
        if lst:
            break

        t_time += 1
    if not lst:
        break
    time = time + t_time
    lst.sort()
    y, x = lst[0]
    arr[y][x] = 0
    r_y, r_x = y, x
    r_exp += 1
    if r_exp == r_size:
        r_size += 1
        r_exp = 0

print(time)
